src/net/socket.rs

```diff
@@ -121,50 +148,85 @@ pub struct Socket {
    rx: mpsc::UnboundedReceiver<relay::Packet>,
 }
 
+type Stream = SplitStream<WebSocketStream<MaybeTlsStream<TcpStream>>>;
+type Sink = SplitSink<WebSocketStream<MaybeTlsStream<TcpStream>>, Message>;
+
 impl Socket {
    async fn read_packet(
-      read_half: &mut tcp::OwnedReadHalf,
-      len: usize,
+      message: tungstenite::Result<Message>,
       output: &mut mpsc::UnboundedSender<relay::Packet>,
-   ) -> anyhow::Result<()> {
-      if len > relay::MAX_PACKET_SIZE as usize {
-         anyhow::bail!("Packet is too big");
+      tx: &broadcast::Sender<Quit>,
+   ) -> anyhow::Result<Option<()>> {
+      match message {
+         Ok(Message::Binary(data)) => {
+            if data.len() > relay::MAX_PACKET_SIZE as usize {
+               anyhow::bail!("Packet is too big");
+            }
+            let packet = bincode::deserialize(&data).context("Invalid packet")?;
+            output.send(packet)?;
+         }
+         Ok(Message::Close(frame)) => {
+            bus::push(Fatal(anyhow::anyhow!("The relay has been disconnected.")));
+
+            if let Some(frame) = frame {
+               log::warn!(
+                  "the relay has been disconnected: {:?}, code: {}",
+                  frame.reason,
+                  frame.code
+               );
+            } else {
+               log::warn!("the relay has been disconnected (reason unknown)");
+            }
+
+            tx.send(Quit)?;
+
+            return Ok(None);
+         }
+         Err(e) => {
+            use tokio_tungstenite::tungstenite::error::ProtocolError;
+            use tokio_tungstenite::tungstenite::Error::*;
+            match e {
+               ConnectionClosed => return Ok(None),
+               // Relay can force a closing handshake, and WebSockets requires a closing handshake.
+               // If we do not get it, it means that relay has been closed and we have to close the session.
+               AlreadyClosed | Protocol(ProtocolError::ResetWithoutClosingHandshake) => {
+                  log::error!("the connection was closed without a closing handshake (relay probably crashed)");
+                  bus::push(Fatal(anyhow::anyhow!("The relay has been disconnected.")));
+                  return Ok(None);
+               }
+               _ => anyhow::bail!(e),
+            }
+         }
+         _ => log::info!("got unused message"),
       }
-      let mut bytes = vec![0; len as usize];
-      read_half.read_exact(&mut bytes).await?;
-      let packet = bincode::deserialize(&bytes).context("Invalid packet")?;
-      output.send(packet)?;
-      Ok(())
+
+      Ok(Some(()))
    }
 
    async fn receiver_loop(
-      mut read_half: tcp::OwnedReadHalf,
+      mut stream: Stream,
       mut output: mpsc::UnboundedSender<relay::Packet>,
-      mut quit: oneshot::Receiver<Quit>,
+      tx: broadcast::Sender<Quit>,
+      mut rx: broadcast::Receiver<Quit>,
```

Comment:
`tx` and `rx` are too generic, please rename them to `quit_tx` and `quit_rx` respectively.